src/mlpack/core/optimizers/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -0,0 +1,125 @@
+/**
+ * @file parallel_sgd_impl.hpp
+ * @author Shikhar Bhardwaj
+ *
+ * Implementation of Parallel Stochastic Gradient Descent.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "parallel_sgd.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template <typename DecayPolicyType>
+ParallelSGD<DecayPolicyType>::ParallelSGD(
+    const size_t maxIterations,
+    const size_t threadShareSize,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy) :
+    maxIterations(maxIterations),
+    threadShareSize(threadShareSize),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy)
+{ /* Nothing to do. */ }
+
+template <typename DecayPolicyType>
+template <typename SparseFunctionType>
+double ParallelSGD<DecayPolicyType>::Optimize(
+    SparseFunctionType& function,
+    arma::mat& iterate)
+{
+  double overallObjective = DBL_MAX;
+  double lastObjective;
+
+  // The order in which the functions will be visited.
+  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+      (function.NumFunctions() - 1), function.NumFunctions());
+
+  // Iterate till the objective is within tolerance or the maximum number of
+  // allowed iterations is reached. If maxIterations is 0, this will iterate
+  // till convergence.
+  for (size_t i = 1; i != maxIterations; ++i)
+  {
+    // Calculate the overall objective.
+    lastObjective = overallObjective;
+    overallObjective = 0;
+
+    for (size_t j = 0; j < function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    // Output current objective function.
+    Log::Info << "Parallel SGD: iteration " << i << ", objective "
+      << overallObjective << "." << std::endl;
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Parallel SGD: converged to " << overallObjective
+        << "; terminating with failure. Try a smaller step size?"
+        << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
+        << "terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    // Shuffle for uniform sampling of functions by each thread.
+    visitationOrder = arma::shuffle(visitationOrder);
```

Comment:
I didn't add one because I initially thought running without shuffling would make no sense. But maybe the user has pre-shuffled data(?) and does not want to incur the cost of a shuffle, while maintaining good locality of reference.